### PR TITLE
Session4

### PR DESCRIPTION
**Pull request by hono-mame**

### CODE REVIEW

File: Session4/PerceptronAndGate.hs [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import Torch.Tensor (Tensor, asTensor, asValue)
```
Comment (bluesky0906) [inline Session4/PerceptronAndGate.hs:1]: I believe this file is nearly identical to Session4/MlpXor_step.hs. To improve code health and avoid duplication, we should consolidate these two files. The nonlinearity could then be passed as an argument from the main function.

File: Session4/README.md [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+  Tensor
+threshold threshold value self =
+  unsafePerformIO $ cast3 ATen.threshold_tss self threshold value
+```
+
+**What I have learned:**  
+It is clear that when using activation functions like **tanh or sigmoid**, higher accuracy can be achieved during training.   
+In contrast, the **step** activation function performs worse compared to the other two.
+Moreover, I noticed that with **tanh, the number of iterations required for the loss to approach zero is smaller than with sigmoid**.  
+This leads me to believe that **tanh is superior to sigmoid.**
```
Comment (bluesky0906) [inline Session4/README.md:332]: Interesting! If you have time, I'd be curious to analyze why tanh is more appropriate than sigmoid for this task. (I'm not sure of the reasons myself, haha.)

File: Session4/README.md [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+number of iterations: 4000
+learning rate: Tensor Float []  0.3000   
+Final Model:
+0, 0 => Tensor Float []  1.1921e-7
+0, 1 => Tensor Float []  1.0000   
+1, 0 => Tensor Float []  1.0000   
+1, 1 => Tensor Float []  5.9605e-8
+---------------------------------------
+```
+![](charts/MlpXor_sigmoid_result.png)
```
Comment (bluesky0906) [inline Session4/README.md:256]: As we covered in the session, here's a reminder:
Let's try with smaller leaning rate.

**Review by bluesky0906 — COMMENTED** [review]
Apologies for the late comment.
Your output was truly great! Keep up the good work!

File: Session4/README.md (hunk 3 above)
Comment (bluesky0906) [reply Session4/README.md:256]: The link might be helpful
https://wandb.ai/mostafaibrahim17/ml-articles/reports/A-Deep-Dive-Into-Learning-Curves-in-Machine-Learning--Vmlldzo0NjA1ODY0
